pkg/dbutil/table.go

```diff
@@ -52,202 +50,30 @@ func GetTableInfoBySQL(createTableSQL string, sqlMode string) (table *model.Tabl
 
 	s, ok := stmt.(*ast.CreateTableStmt)
 	if ok {
-		cols, newConstraints := BuildColumnsAndConstraints(s.Cols, s.Constraints)
-		table, err := BuildTableInfo(s.Table.Name, cols, newConstraints)
+		table, err := ddl.BuildTableInfoFromAST(s)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 
-		return table, nil
-	}
-
-	return nil, errors.Errorf("get table info from sql %s failed!", createTableSQL)
-}
-
-// BuildColumnsAndConstraints reference https://github.com/pingcap/tidb/blob/12c87929b8444571b9e84d2c0d5b85303d27da64/ddl/ddl_api.go#L168
-func BuildColumnsAndConstraints(colDefs []*ast.ColumnDef, constraints []*ast.Constraint) ([]*model.ColumnInfo, []*ast.Constraint) {
-	var cols []*model.ColumnInfo
-	for i, colDef := range colDefs {
-		col, cts := buildColumnAndConstraint(i, colDef)
-		cols = append(cols, col)
-		constraints = append(constraints, cts...)
-	}
-	return cols, constraints
-}
-
-func buildColumnAndConstraint(offset int, colDef *ast.ColumnDef) (*model.ColumnInfo, []*ast.Constraint) {
-	return columnDefToCol(offset, colDef)
-}
-
-// columnDefToCol converts ColumnDef to Col and TableConstraints.
-// now only save the index information for column
-// reference https://github.com/pingcap/tidb/blob/12c87929b8444571b9e84d2c0d5b85303d27da64/ddl/ddl_api.go#L254
-func columnDefToCol(offset int, colDef *ast.ColumnDef) (*model.ColumnInfo, []*ast.Constraint) {
-	var constraints = make([]*ast.Constraint, 0)
-
-	column := &model.ColumnInfo{
-		Offset:    offset,
-		Name:      colDef.Name.Name,
-		FieldType: *colDef.Tp,
-	}
-	col := table.ToColumn(column)
-
-	if colDef.Options != nil {
-		length := types.UnspecifiedLength
-		keys := []*ast.IndexColName{
-			{
-				Column: colDef.Name,
-				Length: length,
-			},
-		}
-
-		for _, v := range colDef.Options {
-			switch v.Tp {
-			case ast.ColumnOptionNotNull:
-				// do nothing
-			case ast.ColumnOptionNull:
-				// do nothing
-			case ast.ColumnOptionAutoIncrement:
-				// do nothing
-			case ast.ColumnOptionPrimaryKey:
-				constraint := &ast.Constraint{Tp: ast.ConstraintPrimaryKey, Keys: keys}
-				constraints = append(constraints, constraint)
-				col.Flag |= mysql.PriKeyFlag
-			case ast.ColumnOptionUniqKey:
-				constraint := &ast.Constraint{Tp: ast.ConstraintUniqKey, Name: colDef.Name.Name.O, Keys: keys}
-				constraints = append(constraints, constraint)
-				col.Flag |= mysql.UniqueKeyFlag
-			case ast.ColumnOptionDefaultValue:
-				// do nothing
-			case ast.ColumnOptionOnUpdate:
-				// do nothing
-			case ast.ColumnOptionComment:
-				// do nothing
-			case ast.ColumnOptionGenerated:
-				// FIXME: use a default string now to make col.IsGenerated() return true, will use real generated expr string later
-				col.GeneratedExprString = "Generated"
-				// do nothing
-			case ast.ColumnOptionFulltext:
-				// do nothing
+		// put primary key in indices
+		if table.PKIsHandle {
+			pkIndex := &model.IndexInfo{
+				Name:    model.NewCIStr("PRIMARY"),
+				Primary: true,
```

Comment:
Do we need to set the `State` to `model.StatePublic`?

Reply:
set in 17be9e7